addon/-private/build-spec-processor.js

```diff
@@ -0,0 +1,26 @@
+import Binding from './binding';
+import ComponentSpec from './component-spec';
+import { transform, rule, simple, rest } from 'botanist';
+
+export default function buildSpecProcessor({ bindKey, componentKey }) {
+  return transform({
+    @rule({ [bindKey]: simple('name') })
+    createBinding({ name }) {
+      if (!name) {
+        throw new Error(`Invalid binding: ""`);
+      } else {
+        return new Binding(name);
+      }
+    },
+
+    @rule({ [componentKey]: simple('name'), ...rest('config') })
+    createComponentSpec({ name, config }, { resolveComponent, owner }) {
+      const path = resolveComponent(name);
+      if (!owner.hasRegistration(`component:${path}`)) {
```

Comment:
@deverstalmage maybe I'm misunderstanding, but isn't that a chicken-and-egg issue? If I already have a component instance, I don't need to validate that it resolves to something in the container.

If we move to a model where the available components are passed in, I guess this question goes away.